feature/preview/src/main/java/com/google/jetpackcamera/feature/preview/ui/RotationUtil.kt

```diff
@@ -0,0 +1,118 @@
+import androidx.compose.foundation.layout.fillMaxSize
+import androidx.compose.runtime.Composable
+import androidx.compose.runtime.DisposableEffect
+import androidx.compose.runtime.LaunchedEffect
+import androidx.compose.runtime.getValue
+import androidx.compose.runtime.mutableIntStateOf
+import androidx.compose.runtime.remember
+import androidx.compose.runtime.setValue
+import androidx.compose.ui.Modifier
+import androidx.compose.ui.layout.layout
+import androidx.compose.ui.platform.LocalConfiguration
+import androidx.compose.ui.platform.LocalView
+import androidx.compose.ui.unit.Constraints
+import androidx.core.view.WindowCompat
+import androidx.core.view.WindowInsetsCompat
+import androidx.core.view.WindowInsetsControllerCompat
+
+/**
+ * As long as this composable is active, the window will go into immersive mode and prevents the
+ * rotation animation on configuration change. This will prevent the UI items from visually changing.
+ *
+ * When used in combination with a composable that renders the same UI in both landscape and portrait,
+ * it can create a smooth continuous feel between those two orientations.
+ */
+@Composable
+fun SmoothImmersiveRotationEffect(context: Context) = DisposableEffect(context) {
+    var currentRotationAnimation: Int? = null
+    context.getActivity()?.window?.let { window ->
+        window.attributes = window.attributes.apply {
+            currentRotationAnimation = rotationAnimation
+            rotationAnimation = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
+                ROTATION_ANIMATION_SEAMLESS
+            } else {
+                ROTATION_ANIMATION_JUMPCUT
+            }
+        }
+        WindowCompat.getInsetsController(window, window.decorView).apply {
+            systemBarsBehavior =
+                WindowInsetsControllerCompat.BEHAVIOR_SHOW_TRANSIENT_BARS_BY_SWIPE
+            hide(WindowInsetsCompat.Type.systemBars())
+        }
+    }
+    onDispose {
+        context.getActivity()?.window?.let { window ->
+            if (currentRotationAnimation != null) {
+                window.attributes = window.attributes.apply {
+                    rotationAnimation = WindowManager.LayoutParams.ROTATION_ANIMATION_ROTATE
+                }
+            }
+            WindowCompat.getInsetsController(window, window.decorView).apply {
+                show(WindowInsetsCompat.Type.systemBars())
+            }
+        }
+    }
+}
+
+/**
+ * Rotate a layout based on the current screen orientation. The UI will always be laid out in a way
+ * that width <= height, and rotated afterwards.
+ */
+@Composable
+fun Modifier.rotatedLayout(): Modifier {
+    var currentOrientation by remember { mutableIntStateOf(Surface.ROTATION_0) }
+    val currentDegrees = currentOrientation * 90f
+    val newOrientation = LocalConfiguration.current.orientation
+    val display = LocalView.current.display
+    LaunchedEffect(newOrientation, display) {
+        val newRotation = display.rotation
```

Comment:
The display rotation can change without `LocalConfiguration.current` changing. I think we could register a [display listener](https://developer.android.com/media/camera/camerax/orientation-rotation#displayListener) to track that?

Reply:
I was able to key off configuration changes to get the updated display rotation. This eliminated an unnecessary recomposition due to the LaunchedEffect modifying MutableState, so we no longer get a flicker in the UI on rotation changes.